File: Python/Reorder List.py
"""
Given a singly linked list L: L0→L1→…→Ln-1→Ln,
reorder it to: L0→Ln→L1→Ln-1→L2→Ln-2→…
You must do this in-place without altering the nodes' values.
For example,
Given {1,2,3,4}, reorder it to {1,4,2,3}.
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @return nothing
    def reorderList(self, head):
        if head is None or head.next is None:
            return head
        
        a,b = self.splitList(head)
        b = self.reverseList(b)
        head = self.mergeList(a,b)
    
    # @param a, a ListNode
    # @param b, a ListNode
    # @return the merged list
    def mergeList(self,a,b):
        head = a
        tail = a
        
        while b:
            tmp = tail.next
            nextNode = b.next
            tail.next = b
            tail.next.next = tmp
            b = nextNode
            tail = tail.next.next
        return head
        
    # @param head,a ListNode
    # @return head, a ListNode which is the first half of head
    # @returm middle, a ListNode which begin from the end of head to the middle of head
    def splitList(self,head):
        if head is None:
            return None
        slow = head
        fast = head
        while fast and fast.next:
            fast = fast.next.next
            slow = slow.next
        middle = slow.next
        slow.next = None
        return head,middle
        
    # return last, a reversed List of head
    def reverseList(self,head):
        if head is None:
            return None
        
        current = head
        last = None
        while current:
            tmp = current.next
            current.next = last
            last = current
            current = tmp
        return last

File: Java/Reorder List.java
/*
Given a singly linked list L: L0→L1→…→Ln-1→Ln,
reorder it to: L0→Ln→L1→Ln-1→L2→Ln-2→…

You must do this in-place without altering the nodes' values.

For example,
Given {1,2,3,4}, reorder it to {1,4,2,3}.
*/


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void reorderList(ListNode head) {
        if(head == null || head.next == null) return;
        
        //get the length
        int length = 0;
        ListNode p = head;
        while(p != null){
            p = p.next;
            length ++;
        }
        
        //find the middle node
        p = head;
        for(int i = 0 ; i < length/2 ; i++){
             p = p.next;
        }
        
        //break into two lists
        ListNode halfHead = p.next;
        p.next = null;
        ListNode newHalfHead = reverse(halfHead);

        //combine two lists
        p = head;
        while(newHalfHead != null){
            ListNode next1 = p.next;
            ListNode next2 = newHalfHead.next;
            p.next = newHalfHead;
            newHalfHead.next = next1;
            p = next1;
            newHalfHead = next2;
        }
    }
    
    private ListNode reverse(ListNode head){
        if(head == null) return head;
        ListNode prev = null;
        ListNode cur = head;
        while(cur != null){
            ListNode next = cur.next;
            cur.next = prev;
            prev = cur;
            cur = next;
        }
        return prev;
    }
}



public class Solution {
    public void reorderList(ListNode head) {
        if(head == null || head.next == null) return;
        
        ListNode prev = null;
        ListNode slow = head;
        ListNode fast = head;
        while(fast != null && fast.next != null){
            prev = slow;
            slow = slow.next;
            fast = fast.next.next;
        }
        prev.next = null;
        slow = reverse(slow);
        
        //combine two lists
        ListNode cur = head;
        while(cur.next != null){
            ListNode next = cur.next;
            cur.next = slow;
            slow = slow.next;
            cur.next.next = next;
            cur = next;
        }
        cur.next = slow;
    }
    
    private ListNode reverse(ListNode head){
        if(head == null) return head;
        ListNode prev = null;
        ListNode cur = head;
        while(cur != null){
            ListNode next = cur.next;
            cur.next = prev;
            prev = cur;
            cur = next;
        }
        return prev;
    }
}